librubyfmt/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+mod de;
+mod delimiters;
+mod file_comments;
+mod format;
+mod intermediary;
+mod line_metadata;
+mod line_tokens;
+mod parser_state;
+mod render_queue_writer;
+mod ripper_tree_types;
+mod ruby;
+mod types;
+
+use file_comments::FileComments;
+use parser_state::ParserState;
+use ruby::VALUE;
+
+#[cfg(debug_assertions)]
+use log::debug;
+#[cfg(debug_assertions)]
+use simplelog::{Config, LevelFilter, TermLogger, TerminalMode};
+
+type RubyfmtResult<T = ()> = std::result::Result<T, Box<dyn std::error::Error>>;
+
+extern "C" {
+    pub fn Init_ripper();
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct FormatBuffer {
+    pub bytes: *const libc::c_char,
+    pub count: i64,
+}
+
+impl FormatBuffer {
+    pub fn into_buf(self) -> &'static [u8] {
+        unsafe { slice::from_raw_parts(self.bytes as *const u8, self.count as usize) }
+    }
+
+    pub fn into_string(self) -> String {
+        unsafe {
+            let vec = Vec::from_raw_parts(
+                self.bytes as *mut u8,
+                self.count as usize,
+                self.count as usize,
+            );
+            String::from_utf8_unchecked(vec)
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum InitStatus {
+    OK = 0,
+    ERROR = 1,
+}
+
+#[no_mangle]
+pub extern "C" fn rubyfmt_init() -> libc::c_int {
+    init_logger();
+    unsafe {
+        ruby::ruby_init();
+    }
+    let res = load_ripper();
+    if res.is_err() {
+        return InitStatus::ERROR as libc::c_int;
+    }
+
+    let res = load_rubyfmt();
+    if res.is_err() {
+        return InitStatus::ERROR as libc::c_int;
+    }
+
+    InitStatus::OK as libc::c_int
+}
+
+pub fn format_buffer(buf: String) -> String {
+    eprintln!("format 1");
+    let bytes: Vec<libc::c_char> = buf
+        .into_bytes()
+        .into_iter()
+        .map(|v| v as libc::c_char)
+        .collect();
+    let len = bytes.len();
+    eprintln!("format 2");
+    let fb = rubyfmt_format_buffer(FormatBuffer {
+        bytes: bytes.as_ptr(),
+        count: len as i64,
+    });
+    eprintln!("format 3");
+    fb.into_string()
+}
+
+#[no_mangle]
+pub extern "C" fn rubyfmt_format_buffer(buf: FormatBuffer) -> FormatBuffer {
+    let output_data = Vec::with_capacity(buf.count as usize);
+    let mut output = Cursor::new(output_data);
+    let tree = run_parser_on(buf);
+    if tree.is_err() {
+        unsafe {
+            let cstr = CString::new("oh no").expect("we just made it");
+            ruby::rb_raise(ruby::rb_eRuntimeError, cstr.as_ptr());
+        }
+    }
+    let tree = tree.expect("we raised");
+    let data = buf.into_buf();
+    let res = toplevel_format_program(&mut output, data, tree);
+    raise_if_error(res);
+    let output_data = output.into_inner().into_boxed_slice();
+    let ptr = output_data.as_ptr();
+    let len = output_data.len();
+    let fb = FormatBuffer {
+        bytes: ptr as *const libc::c_char,
+        count: len as i64,
+    };
+    std::mem::forget(output_data);
+    fb
```

Comment:
I know you're getting rid of `FormatBuffer`, but this is how I'd write this if you were keeping it. Also worth noting that this leaks, and I'd consider including the capacity on the struct to avoid the realloc here.

```suggestion
    output.get_mut().shrink_to_fit();
    let (ptr, len, _) = output.into_inner().into_raw_parts();
    FormatBuffer {
        bytes: ptr as *const libc::c_char,
        count: len as i64,
    }
```